blockchain/message_pool/src/msg_chain.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::{get_gas_perf, get_gas_reward};
+use message::{Message, SignedMessage};
+use num_bigint::BigInt;
+
+/// Represents a node in the MsgChain.
+#[derive(Clone, Debug)]
+pub(crate) struct MsgChainNode {
+    pub msgs: Vec<SignedMessage>,
+    pub gas_reward: BigInt,
+    pub gas_limit: i64,
+    pub gas_perf: f64,
+    pub eff_perf: f64,
+    pub bp: f64,
+    pub parent_offset: f64,
+    pub valid: bool,
+    pub merged: bool,
+}
+
+impl MsgChainNode {
+    pub(crate) fn new() -> Self {
+        Self {
+            msgs: vec![],
+            gas_reward: Default::default(),
+            gas_limit: 0,
+            gas_perf: 0.0,
+            eff_perf: 0.0,
+            bp: 0.0,
+            parent_offset: 0.0,
+            valid: false,
+            merged: false,
+        }
+    }
+    fn set_eff_perf(&mut self, prev: Option<(f64, i64)>) {
+        let mut eff_perf = self.gas_perf * self.bp;
+        if eff_perf > 0.0 && prev.is_some() {
+            let prev = prev.unwrap();
+            let prev_eff_perf = prev.0;
+            let prev_gas_limit = prev.1;
+            let eff_perf_with_parent = (eff_perf * self.gas_limit as f64
+                + prev_eff_perf * prev_gas_limit as f64)
+                / (self.gas_limit + prev_gas_limit) as f64;
+            self.parent_offset = eff_perf - eff_perf_with_parent;
+            eff_perf = eff_perf_with_parent;
+        }
+        self.eff_perf = eff_perf;
+    }
+}
+/// Mimics the doubly linked circular-referenced message chain from Lotus by keeping a current index
+#[derive(Clone, Debug)]
+pub(crate) struct MsgChain {
+    pub index: usize,
+    pub chain: Vec<MsgChainNode>,
+}
+
+impl MsgChain {
```

Comment:
Thats pretty dope didn't realize you could separate the getters and simple commands from  the more complicated logic